File: assets/reports-model/R/Donor Location Distribution.R
# ============================================
# Donor Location Distribution Interactive Chart
# Top 10 Cities/Municipalities
# ============================================

set.seed(123)  # For reproducible sample data
# Sample locations - mostly Iloilo cities and municipalities
sample_locations <- c(
  rep("Iloilo City", 245),
  rep("Oton", 118),
  rep("Pavia", 102),
  rep("Leganes", 89),
  rep("San Miguel", 76),
  rep("Cabatuan", 68),
  rep("Sta. Barbara", 54),
  rep("Tigbauan", 47),
  rep("Maasin", 42),
  rep("Leon", 38),
  rep("Janiuay", 32),
  rep("Badiangan", 28),
  rep("Dumangas", 25),
  rep("Guimbal", 22),
  rep("Alimodian", 18)
)

# Shuffle for randomness
sample_locations <- sample(sample_locations)

# Create a data frame (this simulates your donor_form data)
donor_data_raw <- data.frame(
  donor_id = 1:length(sample_locations),
  location = sample_locations
)

# ============================================
# TO USE YOUR ACTUAL DATA:
# Read your data from CSV or other source
# ============================================
# Option 1: If you have a CSV file with extracted location
# donor_data_raw <- read.csv("donor_form.csv")

# Option 2: If you need to extract city/municipality from permanent_address
# Assuming permanent_address format includes city/municipality
# donor_data_raw <- read.csv("donor_form.csv") %>%
#   mutate(location = sub(".*,\\s*([^,]+),\\s*Iloilo.*", "\\1", permanent_address))

# Option 3: If you have an Excel file
# install.packages("readxl")
# library(readxl)
# donor_data_raw <- read_excel("donor_form.xlsx")

# ============================================
# Data Processing - Get Top 10 Locations
# ============================================
donor_data <- donor_data_raw %>%
  group_by(location) %>%
  summarise(count = n()) %>%
  mutate(
    percentage = round((count / sum(count)) * 100, 1)
  ) %>%
  arrange(desc(count)) %>%
  top_n(10, count)  # Get top 10 only

# Total donors in top 10 locations
total_top10 <- sum(donor_data$count)
total_all <- nrow(donor_data_raw)

# ============================================
# Create Interactive Bar Chart
# ============================================
# Define color gradient - shades of teal/green
colors <- c('#0d9488', '#14b8a6', '#2dd4bf', '#5eead4', '#99f6e4',
            '#0f766e', '#115e59', '#134e4a', '#0d9488', '#14b8a6')

# Create the plot
fig <- plot_ly(
  data = donor_data,
  x = ~reorder(location, -count),
  y = ~count,
  type = 'bar',
  orientation = 'v',
  marker = list(
    color = colors[1:nrow(donor_data)],
    line = list(color = 'rgb(8,51,68)', width = 1.5)
  ),
  hovertext = ~paste0(
    'Location: ', location, '<br>',
    'Count: ', count, '<br>',
    'Percentage: ', percentage, '%'
  ),
  hoverinfo = 'text',
  textposition = 'none'
) %>%
  layout(
    title = list(
      text = paste0(
        '<b>Top 10 Donor Location Distribution</b><br>',
        '<sub>Showing ', total_top10, ' out of ', total_all, ' total donors</sub>'
      ),
      font = list(size = 20)
    ),
    xaxis = list(
      title = '<b>City/Municipality</b>',
      tickfont = list(size = 11),
      titlefont = list(size = 14),
      tickangle = -45
    ),
    yaxis = list(
      title = '<b>Number of Donors</b>',
      tickfont = list(size = 12),
      titlefont = list(size = 14),
      gridcolor = '#e5e7eb'
    ),
    plot_bgcolor = '#f8fafc',
    paper_bgcolor = '#ffffff',
    hovermode = 'closest',
    hoverlabel = list(
      bgcolor = 'white',
      font = list(size = 12, color = 'black'),
      bordercolor = '#e5e7eb'
    ),
    margin = list(t = 100, b = 120, l = 60, r = 30)
  ) %>%
  config(
    displayModeBar = TRUE,
    modeBarButtonsToRemove = c('lasso2d', 'select2d'),
    displaylogo = FALSE
  )

# Display the chart
fig

# ============================================
# Print Summary Statistics
# ============================================
cat("\n=== TOP 10 DONOR LOCATION DISTRIBUTION ===\n\n")
cat(sprintf("Total Donors (All Locations): %d\n", total_all))
cat(sprintf("Total Donors (Top 10): %d\n", total_top10))
cat(sprintf("Coverage: %.1f%%\n\n", (total_top10/total_all)*100))

cat("Rank | Location              | Count | Percentage\n")
cat("-----|----------------------|-------|------------\n")

for(i in 1:nrow(donor_data)) {
  cat(sprintf(
    "%-4d | %-20s | %-5d | %.1f%%\n",
    i,
    donor_data$location[i],
    donor_data$count[i],
    donor_data$percentage[i]
  ))
}

# ============================================
# Additional Analysis - Urban vs Rural
# ============================================
# Classify Iloilo City as urban, others as rural/suburban
location_type <- donor_data_raw %>%
  mutate(
    area_type = ifelse(location == "Iloilo City", "Urban", "Rural/Suburban")
  ) %>%
  group_by(area_type) %>%
  summarise(count = n()) %>%
  mutate(percentage = round((count / sum(count)) * 100, 1))

cat("\n=== AREA TYPE DISTRIBUTION ===\n\n")
for(i in 1:nrow(location_type)) {
  cat(sprintf(
    "%s: %d donors (%.1f%%)\n",
    location_type$area_type[i],
    location_type$count[i],
    location_type$percentage[i]
  ))
}

File: assets/reports-model/Donor Location Distribution.py
"""
Donor Location Distribution Interactive Chart
==============================================
Creates an interactive bar chart showing top 10 donor locations.
Uses ONLY donors whose latest eligibility status is approved/eligible, so the
total donors for this chart matches the active donor universe.
"""

import json
import re
import sys
from typing import Dict, List, Set

import plotly.graph_objects as go

from database import DatabaseConnection


# Color gradient - shades of teal/green (matching R version)
COLORS = ["#0d9488", "#14b8a6", "#2dd4bf", "#5eead4", "#99f6e4",
          "#0f766e", "#115e59", "#134e4a", "#0d9488", "#14b8a6"]


def _parse_dt(value):
    from datetime import datetime

    if not value:
        return None
    try:
        text = str(value)
        if "Z" in text:
            return datetime.fromisoformat(text.replace("Z", "+00:00"))
        return datetime.fromisoformat(text)
    except Exception:
        return None


def fetch_active_donor_ids() -> Set[int]:
    """
    Fetch unique donor_ids whose LATEST eligibility status is APPROVED/eligible.
    Shared definition with Total Active Donors and other demographic charts.
    """
    db = DatabaseConnection()
    records: List[Dict] = []

    if db.connect():
        try:
            endpoint = (
                "eligibility?"
                "select=eligibility_id,donor_id,status,created_at"
                "&order=donor_id.asc,created_at.asc"
            )
            records = db.supabase_request(endpoint)
        finally:
            db.disconnect()

    latest_by_donor: Dict[int, Dict] = {}
    for rec in records:
        donor_id = rec.get("donor_id")
        if donor_id is None:
            continue
        created = _parse_dt(rec.get("created_at"))
        existing = latest_by_donor.get(donor_id)
        if existing is None:
            latest_by_donor[donor_id] = {"status": rec.get("status"), "_created": created}
        else:
            prev_created = existing.get("_created")
            if created and (prev_created is None or created > prev_created):
                latest_by_donor[donor_id] = {"status": rec.get("status"), "_created": created}

    active_ids: Set[int] = set()
    for donor_id, info in latest_by_donor.items():
        status_text = (info.get("status") or "").lower()
        if "approved" in status_text or status_text == "eligible":
            active_ids.add(donor_id)

    return active_ids


def fetch_donor_forms(filtered_ids: Set[int]) -> List[Dict]:
    """Fetch donor forms with address, limited to donor_ids in filtered_ids."""
    if not filtered_ids:
        return []

    db = DatabaseConnection()
    donor_forms: List[Dict] = []

    if db.connect():
        try:
            endpoint = (
                "donor_form?"
                "select=donor_id,permanent_address"
            )
            rows = db.supabase_request(endpoint)
        finally:
            db.disconnect()

    allowed = set(filtered_ids)
    for row in rows:
        donor_id = row.get("donor_id")
        if donor_id in allowed:
            donor_forms.append(row)

    return donor_forms


def extract_location(address: str) -> str:
    """
    Extract city/municipality from permanent address.
    Handles various Philippine address formats.
    """
    if not address:
        return "Unknown"
    
    address = address.strip()
    
    # Common patterns for Philippine addresses
    # Try to extract the municipality/city (usually before the province)
    
    # Pattern 1: "..., City/Municipality, Province"
    # Pattern 2: "Barangay X, City/Municipality, Province"
    
    # Split by comma and analyze parts
    parts = [p.strip() for p in address.split(",")]
    
    if len(parts) >= 2:
        # Look for city/municipality keywords
        for i, part in enumerate(parts):
            part_lower = part.lower()
            # Check if this part looks like a city/municipality
            if "city" in part_lower:
                return part.strip()
            # Check for common Iloilo municipalities
            municipalities = [
                "oton", "pavia", "leganes", "san miguel", "cabatuan",
                "sta. barbara", "santa barbara", "tigbauan", "maasin", "leon",
                "janiuay", "badiangan", "dumangas", "guimbal", "alimodian",
                "miagao", "san joaquin", "anilao", "banate", "barotac nuevo",
                "barotac viejo", "batad", "bingawan", "calinog", "carles",
                "concepcion", "dingle", "dueñas", "estancia", "igbaras",
                "lambunao", "lemery", "mina", "new lucena", "passi",
                "pototan", "san dionisio", "san enrique", "san rafael",
                "sara", "zarraga"
            ]
            for muni in municipalities:
                if muni in part_lower:
                    return part.strip()
        
        # If no municipality found, try second-to-last part (common format)
        if len(parts) >= 3:
            return parts[-2].strip()
        else:
            return parts[-1].strip()
    
    # Return first significant part if can't parse
    return parts[0] if parts else "Unknown"


def aggregate_by_location(donor_forms: List[Dict], top_n: int = 10) -> List[Dict]:
    """
    Aggregate donors by location (city/municipality).
    Returns top N locations sorted by count.
    """
    location_counts: Dict[str, int] = {}
    
    for donor in donor_forms:
        address = donor.get("permanent_address", "")
        location = extract_location(address)
        
        # Normalize location name
        location = location.title() if location else "Unknown"
        
        location_counts[location] = location_counts.get(location, 0) + 1
    
    # Sort by count and get top N
    sorted_locations = sorted(location_counts.items(), key=lambda x: x[1], reverse=True)
    top_locations = sorted_locations[:top_n]
    
    total_all = sum(location_counts.values())
    
    result = []
    for location, count in top_locations:
        percentage = round((count / total_all) * 100, 1) if total_all > 0 else 0
        result.append({
            "location": location,
            "count": count,
            "percentage": percentage
        })
    
    return result


def calculate_area_type_summary(donor_forms: List[Dict], urban_locations: List[str] = None) -> List[Dict]:
    """Calculate Urban vs Rural/Suburban summary."""
    if urban_locations is None:
        urban_locations = ["Iloilo City", "Iloilo"]
    
    urban_count = 0
    rural_count = 0
    
    for donor in donor_forms:
        address = donor.get("permanent_address", "")
        location = extract_location(address)
        
        if any(urban.lower() in location.lower() for urban in urban_locations):
            urban_count += 1
        else:
            rural_count += 1
    
    total = urban_count + rural_count
    
    return [
        {
            "area_type": "Urban",
            "count": urban_count,
            "percentage": round((urban_count / total) * 100, 1) if total > 0 else 0
        },
        {
            "area_type": "Rural/Suburban",
            "count": rural_count,
            "percentage": round((rural_count / total) * 100, 1) if total > 0 else 0
        }
    ]


def create_bar_chart(location_data: List[Dict], total_top: int, total_all: int) -> go.Figure:
    """Create an interactive bar chart using Plotly."""
    locations = [d["location"] for d in location_data]
    counts = [d["count"] for d in location_data]
    
    fig = go.Figure(data=[
        go.Bar(
            x=locations,
            y=counts,
            marker=dict(
                color=COLORS[:len(locations)],
                line=dict(color="rgb(8,51,68)", width=1.5)
            ),
            hovertemplate=(
                "Location: %{x}<br>"
                "Count: %{y}<br>"
                "<extra></extra>"
            )
        )
    ])
    
    fig.update_layout(
        title=dict(
            text=f"<b>Top 10 Donor Location Distribution</b><br><sub>Showing {total_top} out of {total_all} total donors</sub>",
            font=dict(size=20)
        ),
        xaxis=dict(
            title="<b>City/Municipality</b>",
            tickfont=dict(size=11),
            tickangle=-45
        ),
        yaxis=dict(
            title="<b>Number of Donors</b>",
            tickfont=dict(size=12),
            gridcolor="#e5e7eb"
        ),
        plot_bgcolor="#f8fafc",
        paper_bgcolor="#ffffff",
        hovermode="closest",
        hoverlabel=dict(
            bgcolor="white",
            font_size=12,
            font_color="black",
            bordercolor="#e5e7eb"
        ),
        margin=dict(t=100, b=120, l=60, r=30)
    )
    
    return fig


def get_donor_location_distribution_data(top_n: int = 10) -> Dict:
    """
    Main function to get donor location distribution data.
    Returns a dictionary with the data for API consumption.
    """
    donor_ids = fetch_active_donor_ids()
    donor_forms = fetch_donor_forms(donor_ids)

    if not donor_forms:
        return {
            "success": False,
            "message": "No active donors with location data found in database",
            "data": [],
            "total_top": 0,
            "total_all": 0,
            "area_type_summary": [],
        }

    location_data = aggregate_by_location(donor_forms, top_n)
    total_top = sum(d["count"] for d in location_data)
    total_all = len(donor_forms)
    area_type_summary = calculate_area_type_summary(donor_forms)
    
    return {
        "success": True,
        "data": location_data,
        "total_top": total_top,
        "total_all": total_all,
        "coverage_percentage": round((total_top / total_all) * 100, 1) if total_all > 0 else 0,
        "area_type_summary": area_type_summary
    }


def generate_chart_html(output_path: str = None) -> str:
    """Generate and optionally save the chart as HTML.

    IMPORTANT: Always create an HTML asset file, even when there is no data,
    so that forecast-asset.php never returns a 404. In the no-data case,
    we write a simple placeholder HTML with a friendly message.
    """
    try:
        donor_ids = fetch_active_donor_ids()
        donor_forms = fetch_donor_forms(donor_ids)

        if not donor_forms:
            placeholder_html = """<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Donor Location Distribution</title></head>
<body style="font-family: Arial, sans-serif; padding: 16px;">
  <h3 style="margin-top: 0;">Donor Location Distribution</h3>
  <p>No active donors with location data are currently available. Please check back later.</p>
</body></html>"""
            if output_path:
                with open(output_path, "w", encoding="utf-8") as f:
                    f.write(placeholder_html)
                print(f"Placeholder location chart saved to: {output_path}", file=sys.stderr)
            return placeholder_html

        location_data = aggregate_by_location(donor_forms, 10)
        total_top = sum(d["count"] for d in location_data)
        total_all = len(donor_forms)

        fig = create_bar_chart(location_data, total_top, total_all)

        if output_path:
            fig.write_html(output_path)
            print(f"Chart saved to: {output_path}", file=sys.stderr)

        return fig.to_html(include_plotlyjs="cdn")
    except Exception as exc:
        placeholder_html = f"""<!DOCTYPE html>
<html><head><meta charset="utf-8"><title>Donor Location Distribution</title></head>
<body style="font-family: Arial, sans-serif; padding: 16px;">
  <h3 style="margin-top: 0;">Donor Location Distribution</h3>
  <p>An error occurred while generating this chart. Please check server logs.</p>
  <pre style="font-size: 11px; color: #6b7280;">{str(exc)}</pre>
</body></html>"""
        if output_path:
            try:
                with open(output_path, "w", encoding="utf-8") as f:
                    f.write(placeholder_html)
                print(f"Error placeholder location chart saved to: {output_path}", file=sys.stderr)
            except Exception:
                pass
        return placeholder_html


def print_summary():
    """Print a formatted summary to console (matching R output style)."""
    result = get_donor_location_distribution_data()
    
    print("\n=== TOP 10 DONOR LOCATION DISTRIBUTION ===\n")
    print(f"Total Donors (All Locations): {result['total_all']}")
    print(f"Total Donors (Top 10): {result['total_top']}")
    print(f"Coverage: {result['coverage_percentage']}%\n")
    
    print("Rank | Location              | Count | Percentage")
    print("-----|----------------------|-------|------------")
    
    for i, item in enumerate(result["data"], 1):
        print(f"{i:<4} | {item['location']:<20} | {item['count']:<5} | {item['percentage']}%")
    
    # Area type summary
    print("\n=== AREA TYPE DISTRIBUTION ===\n")
    for item in result["area_type_summary"]:
        print(f"{item['area_type']}: {item['count']} donors ({item['percentage']}%)")


if __name__ == "__main__":
    import argparse
    
    parser = argparse.ArgumentParser(description="Donor Location Distribution Report")
    parser.add_argument("--top", type=int, default=10, help="Number of top locations to show")
    parser.add_argument("--json", action="store_true", help="Output as JSON")
    parser.add_argument("--html", type=str, help="Save chart as HTML to specified path")
    args = parser.parse_args()
    
    if args.html:
        generate_chart_html(args.html)
    elif args.json:
        result = get_donor_location_distribution_data(args.top)
        print(json.dumps(result, indent=2))
    else:
        print_summary()

